modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPkgDialect.java

```diff
@@ -82,6 +85,18 @@ public GeoPkgDialect(JDBCDataStore dataStore) {
         geomWriterConfig = new GeoPkgGeomWriter.Configuration();
     }
 
+    // The JDBC aggregate functions doesn't support the geopkg DateTime types.
+    //  This is because they are stored as strings in the database instead of real Date types
+    //   and the JDBC aggregate driver does NOT provide the needed conversions.
+    //
+    // since this doesn't work, we don't support it.
+    @Override
+    public void registerAggregateFunctions(
```

Comment:
Ouch, this fixes the issue but severely limits usability of geopackages with time/elevation/custom dimensions, when having fast min/max/unique is a must. Production systems will work only on very small geopackages in this case.

Thinking out loud, if it's really just a matter of conversion of result datatype, one might consider also doing type conversions around here?
https://github.com/geotools/geotools/blob/bae4ea158dd39cf766f7ead19c93a81222a4cce5/modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java#L1499
Or is it sqlite that does the aggregate in an unexpected way? min/max should work even with strings, aren't dates are saved as ISO dates in sqlite, so they should sort lexicographically anyways?

Just a thought.

Reply:
Yes - there's a line all setup for the conversion here;
https://github.com/geotools/geotools/blob/bae4ea158dd39cf766f7ead19c93a81222a4cce5/modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java#L1502

However, there isn't much support for the conversion, so it's likely going to require interface changes.  

I ran into this issue in the Time dimension GetCapabilities - which runs a UniqueVisitor on the dataset.  There were several problems in the GeoPKG with this - i.e. having to change the column from "time" to "datetime(time)" in the filters (otherwise the DB returns the wrong results).

Reply:
Right, my only worry is that caps generation will become really really slow... we did quite a bit of work optimizing it with visitors back in the day exactly because caps gen time was getting out of hand real quick...

Reply:
I could not find a way of checking the columns the operation is operating on, like if we can confirm everything is simple (and not subject to mapping) then the visitors can be used. In the specific use of GetCapabilities generation making a summary of TIMESTAMP is a common case.

I had thought about making the CalcResult smarter (giving it a target class that it accepts, and it could be responsible for providing a conversation as values are collected). This would solve the problem "once" and not involve making special cases in all the places where visitors are called.

Reply:
Thinking out loud here... the visitor does not really know what the type is, because it's determined at runtime evaluating expressions (besides the visitors that only work on a specific type). JDBCDataStore on the other hand [has the aggregateExpression](https://github.com/geotools/geotools/blob/d8d38228e57638927b745f31bc0b5623f48e61b9/modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java#L1432), from which one can determine the type of the visitor input by using the ExpressionTypeVisitor. 
Now, that is not enough, many visitors just return the same type as the input, but for example avg and sttdev do not.
It would be easy however to add a default method to FeatureCalc or FeatureAttributeVisitor or to the implementations (getExpression is really not present in FeatureCalc for example, but determined through reflection.. not sure why, since there is FeatureAttributeVisitor that can do the job, probably historical reasons).

```
/**
 * Returns the expected output type given the input type, or null if uknown
Class[] getResultType(Class[] inputs);
*/
```

With that one could do the right type conversion no?

Reply:
Mind, these "thinking out loud" mean "let's talk but I'm not asking you to do it". If you are hard pressed to get a fix in, I'd suggest to go head with this, open a new ticket warning users that performance of aggregations on sqlite will become really bad (for the cases where it works, like numbers and strings, which is used for example by SLDService), and hope that someone steps in to follow up. Right now I don't have an installation of gpkg + aggregations that would be affected, someone out there might though.

Reply:
I just added a geopackage ‘model.gpkg’ with several tables and views in it to 
https://osgeo-org.atlassian.net/browse/GEOS-9392

To test I would use  ‘view3’  (>34000 polygons and a datetime column) with given model.sld.

But to test how it ‘detoriates’ the generation of the capabilities it is also a good one because it contains 3 views and tables.

Reply:
https://osgeo-org.atlassian.net/browse/GEOT-6520

@jodygarnett  - I think this is ok to merge now.

Reply:
Thanks I will review